<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.9.1" />
<title>mdsine2.raw_data.util API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>mdsine2.raw_data.util</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">import pandas as pd
from ete3 import Tree
from os.path import dirname, join
import os

_HEALTHY_SUBJECTS = set([&#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;])
_UC_SUBJECTS = set([&#39;6&#39;, &#39;7&#39;, &#39;8&#39;, &#39;9&#39;, &#39;10&#39;])


def load_phylogenetic_tree(with_reference_sequences):
    &#39;&#39;&#39;Load phylogenetic tree into an `ete3.Tree` object.

    Parameters
    ----------
    with_reference_sequences : bool
        If True, load the phylogenetic tree with the reference sequeunces from
        placement. If False, only include the ASVs and not the reference 
        sequences

    Returns
    -------
    ete3.Tree
    &#39;&#39;&#39;
    module_path = dirname(__file__)
    if with_reference_sequences:
        path = join(dirname(__file__),&#39;gibson_dataset&#39; ,&#39;phylogenetic_tree_with_reference.nhx&#39;)
    else:
        path = join(dirname(__file__),&#39;gibson_dataset&#39; ,&#39;phylogenetic_tree_w_branch_len_preserved.nhx&#39;)

    return Tree(path)

def load_qpcr_masses(healthy=None):
    &#39;&#39;&#39;Load qpcr masses table into a `pandas.DataFrame`.

    Parameters
    ----------
    healthy : bool, None
        Whether or not to return the healthy cohort. 
        If True only return the healthy cohort.
        If False only return the ulcerative colitis cohort.
        If None then return everything

    Returns
    -------
    pandas.DataFrame
    &#39;&#39;&#39;
    path = join(dirname(__file__),&#39;gibson_dataset&#39; ,&#39;qpcr.tsv&#39;)
    df = pd.read_csv(path, sep=&#39;\t&#39;)
    df = df.set_index(&#39;sampleID&#39;)

    if healthy is not None:
        keep = []

        for sampleid in df.index:
            if healthy and _is_healthy_sampleid(sampleid):
                keep.append(sampleid)
            elif (not healthy) and (not _is_healthy_sampleid(sampleid)):
                keep.append(sampleid)

        df = df.loc[keep]
    return df

def load_reads(healthy=None):
    &#39;&#39;&#39;Load reads table into a `pandas.DataFrame`.

    Parameters
    ----------
    healthy : bool, None
        Whether or not to return the healthy cohort. 
        If True only return the healthy cohort.
        If False only return the ulcerative colitis cohort.
        If None then return everything

    Returns
    -------
    pandas.DataFrame
    &#39;&#39;&#39;
    path = join(dirname(__file__),&#39;gibson_dataset&#39; ,&#39;reads.tsv&#39;)
    df = pd.read_csv(path, sep=&#39;\t&#39;)
    df = df.set_index(&#39;name&#39;)

    if healthy is not None:
        keep = []

        for sampleid in df.columns:
            if healthy and _is_healthy_sampleid(sampleid):
                keep.append(sampleid)
            elif (not healthy) and (not _is_healthy_sampleid(sampleid)):
                keep.append(sampleid)

        df = df[keep]
    return df

def load_sampleid(healthy=None):
    &#39;&#39;&#39;Load sample ID table into a `pandas.DataFrame`.

    Parameters
    ----------
    healthy : bool, None
        Whether or not to return the healthy cohort. 
        If True only return the healthy cohort.
        If False only return the ulcerative colitis cohort.
        If None then return everything

    Returns
    -------
    pandas.DataFrame
    &#39;&#39;&#39;
    path = join(dirname(__file__),&#39;gibson_dataset&#39; ,&#39;sampleid.tsv&#39;)
    df = pd.read_csv(path, sep=&#39;\t&#39;)
    df = df.set_index(&#39;sampleID&#39;)

    if healthy is not None:
        keep = []

        for sampleid in df.index:
            if healthy and _is_healthy_sampleid(sampleid):
                keep.append(sampleid)
            elif (not healthy) and (not _is_healthy_sampleid(sampleid)):
                keep.append(sampleid)

        df = df.loc[keep]
    return df

def load_taxonomy():
    &#39;&#39;&#39;Load taxonomy table into a `pandas.DataFrame`.

    Returns
    -------
    pandas.DataFrame
    &#39;&#39;&#39;
    path = join(dirname(__file__),&#39;gibson_dataset&#39; ,&#39;taxonomy.tsv&#39;)
    return pd.read_csv(path, sep=&#39;\t&#39;)

def _is_healthy_sampleid(sampleid):
    &#39;&#39;&#39;Checks whether the sample id is for healty or UC

    Parameters
    ----------
    sampleid : str
        Sample ID

    Returns
    -------
    bool
    &#39;&#39;&#39;
    sid = sampleid.split(&#39;-&#39;)[0]
    return sid in _HEALTHY_SUBJECTS</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="mdsine2.raw_data.util.load_phylogenetic_tree"><code class="name flex">
<span>def <span class="ident">load_phylogenetic_tree</span></span>(<span>with_reference_sequences)</span>
</code></dt>
<dd>
<div class="desc"><p>Load phylogenetic tree into an <code>ete3.Tree</code> object.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>with_reference_sequences</code></strong> :&ensp;<code>bool</code></dt>
<dd>If True, load the phylogenetic tree with the reference sequeunces from
placement. If False, only include the ASVs and not the reference
sequences</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>ete3.Tree</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def load_phylogenetic_tree(with_reference_sequences):
    &#39;&#39;&#39;Load phylogenetic tree into an `ete3.Tree` object.

    Parameters
    ----------
    with_reference_sequences : bool
        If True, load the phylogenetic tree with the reference sequeunces from
        placement. If False, only include the ASVs and not the reference 
        sequences

    Returns
    -------
    ete3.Tree
    &#39;&#39;&#39;
    module_path = dirname(__file__)
    if with_reference_sequences:
        path = join(dirname(__file__),&#39;gibson_dataset&#39; ,&#39;phylogenetic_tree_with_reference.nhx&#39;)
    else:
        path = join(dirname(__file__),&#39;gibson_dataset&#39; ,&#39;phylogenetic_tree_w_branch_len_preserved.nhx&#39;)

    return Tree(path)</code></pre>
</details>
</dd>
<dt id="mdsine2.raw_data.util.load_qpcr_masses"><code class="name flex">
<span>def <span class="ident">load_qpcr_masses</span></span>(<span>healthy=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Load qpcr masses table into a <code>pandas.DataFrame</code>.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>healthy</code></strong> :&ensp;<code>bool, None</code></dt>
<dd>Whether or not to return the healthy cohort.
If True only return the healthy cohort.
If False only return the ulcerative colitis cohort.
If None then return everything</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>pandas.DataFrame</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def load_qpcr_masses(healthy=None):
    &#39;&#39;&#39;Load qpcr masses table into a `pandas.DataFrame`.

    Parameters
    ----------
    healthy : bool, None
        Whether or not to return the healthy cohort. 
        If True only return the healthy cohort.
        If False only return the ulcerative colitis cohort.
        If None then return everything

    Returns
    -------
    pandas.DataFrame
    &#39;&#39;&#39;
    path = join(dirname(__file__),&#39;gibson_dataset&#39; ,&#39;qpcr.tsv&#39;)
    df = pd.read_csv(path, sep=&#39;\t&#39;)
    df = df.set_index(&#39;sampleID&#39;)

    if healthy is not None:
        keep = []

        for sampleid in df.index:
            if healthy and _is_healthy_sampleid(sampleid):
                keep.append(sampleid)
            elif (not healthy) and (not _is_healthy_sampleid(sampleid)):
                keep.append(sampleid)

        df = df.loc[keep]
    return df</code></pre>
</details>
</dd>
<dt id="mdsine2.raw_data.util.load_reads"><code class="name flex">
<span>def <span class="ident">load_reads</span></span>(<span>healthy=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Load reads table into a <code>pandas.DataFrame</code>.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>healthy</code></strong> :&ensp;<code>bool, None</code></dt>
<dd>Whether or not to return the healthy cohort.
If True only return the healthy cohort.
If False only return the ulcerative colitis cohort.
If None then return everything</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>pandas.DataFrame</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def load_reads(healthy=None):
    &#39;&#39;&#39;Load reads table into a `pandas.DataFrame`.

    Parameters
    ----------
    healthy : bool, None
        Whether or not to return the healthy cohort. 
        If True only return the healthy cohort.
        If False only return the ulcerative colitis cohort.
        If None then return everything

    Returns
    -------
    pandas.DataFrame
    &#39;&#39;&#39;
    path = join(dirname(__file__),&#39;gibson_dataset&#39; ,&#39;reads.tsv&#39;)
    df = pd.read_csv(path, sep=&#39;\t&#39;)
    df = df.set_index(&#39;name&#39;)

    if healthy is not None:
        keep = []

        for sampleid in df.columns:
            if healthy and _is_healthy_sampleid(sampleid):
                keep.append(sampleid)
            elif (not healthy) and (not _is_healthy_sampleid(sampleid)):
                keep.append(sampleid)

        df = df[keep]
    return df</code></pre>
</details>
</dd>
<dt id="mdsine2.raw_data.util.load_sampleid"><code class="name flex">
<span>def <span class="ident">load_sampleid</span></span>(<span>healthy=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Load sample ID table into a <code>pandas.DataFrame</code>.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>healthy</code></strong> :&ensp;<code>bool, None</code></dt>
<dd>Whether or not to return the healthy cohort.
If True only return the healthy cohort.
If False only return the ulcerative colitis cohort.
If None then return everything</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>pandas.DataFrame</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def load_sampleid(healthy=None):
    &#39;&#39;&#39;Load sample ID table into a `pandas.DataFrame`.

    Parameters
    ----------
    healthy : bool, None
        Whether or not to return the healthy cohort. 
        If True only return the healthy cohort.
        If False only return the ulcerative colitis cohort.
        If None then return everything

    Returns
    -------
    pandas.DataFrame
    &#39;&#39;&#39;
    path = join(dirname(__file__),&#39;gibson_dataset&#39; ,&#39;sampleid.tsv&#39;)
    df = pd.read_csv(path, sep=&#39;\t&#39;)
    df = df.set_index(&#39;sampleID&#39;)

    if healthy is not None:
        keep = []

        for sampleid in df.index:
            if healthy and _is_healthy_sampleid(sampleid):
                keep.append(sampleid)
            elif (not healthy) and (not _is_healthy_sampleid(sampleid)):
                keep.append(sampleid)

        df = df.loc[keep]
    return df</code></pre>
</details>
</dd>
<dt id="mdsine2.raw_data.util.load_taxonomy"><code class="name flex">
<span>def <span class="ident">load_taxonomy</span></span>(<span>)</span>
</code></dt>
<dd>
<div class="desc"><p>Load taxonomy table into a <code>pandas.DataFrame</code>.</p>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>pandas.DataFrame</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def load_taxonomy():
    &#39;&#39;&#39;Load taxonomy table into a `pandas.DataFrame`.

    Returns
    -------
    pandas.DataFrame
    &#39;&#39;&#39;
    path = join(dirname(__file__),&#39;gibson_dataset&#39; ,&#39;taxonomy.tsv&#39;)
    return pd.read_csv(path, sep=&#39;\t&#39;)</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="mdsine2.raw_data" href="index.html">mdsine2.raw_data</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="mdsine2.raw_data.util.load_phylogenetic_tree" href="#mdsine2.raw_data.util.load_phylogenetic_tree">load_phylogenetic_tree</a></code></li>
<li><code><a title="mdsine2.raw_data.util.load_qpcr_masses" href="#mdsine2.raw_data.util.load_qpcr_masses">load_qpcr_masses</a></code></li>
<li><code><a title="mdsine2.raw_data.util.load_reads" href="#mdsine2.raw_data.util.load_reads">load_reads</a></code></li>
<li><code><a title="mdsine2.raw_data.util.load_sampleid" href="#mdsine2.raw_data.util.load_sampleid">load_sampleid</a></code></li>
<li><code><a title="mdsine2.raw_data.util.load_taxonomy" href="#mdsine2.raw_data.util.load_taxonomy">load_taxonomy</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.9.1</a>.</p>
</footer>
</body>
</html>